docs/devel/api-conventions.md

```diff
@@ -247,6 +248,14 @@ ports:
 
 This rule maintains the invariant that all JSON/YAML keys are fields in API objects. The only exceptions are pure maps in the API (currently, labels, selectors, annotations, data), as opposed to sets of subobjects.
 
+#### Primitive types
+
+* Avoid floating-point values as much as possible, and never use them in spec. Floating-point values cannot be reliably round-tripped (encoded and re-decoded) without changing, and have varying precision and representations across languages and architectures.
+* All numbers (e.g., uint32, int64) are converted to float64 by Javascript and some other languages, so any field which is expected to exceed that either in magnitude or in precision (specifically integer values > 53 bits) should be serialized and accepted as strings.
+* Do not use unsigned integers, due to inconsistent support across languages and libraries. Just validate that the integer is non-negative if that's the case.
```

Comment:
I could be ok with this - @thockin if you agree I'll use int64 internally
for UID/GID and then we'll validate to the range

On Wed, Nov 18, 2015 at 1:05 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In docs/devel/api-conventions.md
> <https://github.com/kubernetes/kubernetes/pull/17377#discussion_r45235687>
> :
>
> > @@ -247,6 +248,14 @@ ports:
> >
> >  This rule maintains the invariant that all JSON/YAML keys are fields in API objects. The only exceptions are pure maps in the API (currently, labels, selectors, annotations, data), as opposed to sets of subobjects.
> >
> > +#### Primitive types
> > +
> > +* Avoid floating-point values as much as possible, and never use them in spec. Floating-point values cannot be reliably round-tripped (encoded and re-decoded) without changing, and have varying precision and representations across languages and architectures.
> > +* All numbers (e.g., uint32, int64) are converted to float64 by Javascript and some other languages, so any field which is expected to exceed that either in magnitude or in precision (specifically integer values > 53 bits) should be serialized and accepted as strings.
> > +* Do not use unsigned integers, due to inconsistent support across languages and libraries. Just validate that the integer is non-negative if that's the case.
>
> @smarterclayton <https://github.com/smarterclayton> I sort of like
> this... int64 validated to be 0 <= x <= 2^32 instead of uint32? what do
> you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17377/files#r45235687>.
>